File: code/17-Letter Combinations of a Phone Number/17-Letter Combinations of a Phone Number.py
# Given a string containing digits from 2-9 inclusive, return all possible letter combinations that the number could represent.
#
# A mapping of digit to letters (just like on the telephone buttons) is given below. Note that 1 does not map to any letters.
#
# Example:
#
# Input: "23"
# Output: ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"].

class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        phone = {
            '2': ['a', 'b', 'c'],
            '3': ['d', 'e', 'f'],
            '4': ['g', 'h', 'i'],
            '5': ['j', 'k', 'l'],
            '6': ['m', 'n', 'o'],
            '7': ['p', 'q', 'r', 's'],
            '8': ['t', 'u', 'v'],
            '9': ['w', 'x', 'y', 'z']
        }

        output = []

        def backtrack(combination, next_digit):
            if len(next_digit) == 0:
                return output.append(combination)
            else:
                for letter in phone[next_digit[0]]:
                    backtrack(combination + letter, next_digit[1:])

        if digits:
            backtrack('', digits)
        return output

File: code/17-Letter Combinations of a Phone Number/17-Letter Combinations of a Phone Number.cpp
/*
LeetCode 17 - Letter Combinations of a Phone Number

解题步骤（自动整理）
1. 用 DFS/回溯枚举所有可能选择，并维护当前路径/状态
2. 在递归过程中进行剪枝（如边界、重复、提前失败条件）
3. 到达终止条件时收集结果或返回，并回溯撤销选择
*/
/*Letter Combinations of a Phone Number

My Submissions Question Solution
Given a digit string, return all possible letter combinations that the number could represent.

A mapping of digit to letters (just like on the telephone buttons) is given below.



Input:Digit string "23"
Output: ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"].
Note:
Although the above answer is in lexicographical order, your answer could be in any order you want.
*/

class Solution {
public:

    const vector<string> keyword {"","","abc","def","ghi","jkl","mno", "pqrs","tuv","wxyz"};

    void dfs(string &digits, size_t cur, string path, vector<string> &result){

        if(cur == digits.size()){
            result.push_back(path);
            return ;
        }

        for(char c: keyword[digits[cur] - '0'])
           dfs(digits, cur + 1, path + c, result);
    }

    vector<string> letterCombinations(string digits) {

        vector<string> result;

        if(digits.size() == 0)
          return result;

        dfs(digits, 0, "", result);

        return result;

    }


};
